glean/src/core/metrics/types/url.ts

```diff
@@ -0,0 +1,146 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import type { CommonMetricData } from "../index.js";
+import { isString } from "../../utils.js";
+import { MetricType } from "../index.js";
+import { Context } from "../../context.js";
+import { Metric } from "../metric.js";
+import { ErrorType } from "../../error/error_type.js";
+
+// The maximum number of characters a URL Metric may have, before encoding.
+const URL_MAX_LENGTH = 2048;
+
+// This regex only validates that the `scheme` part of the URL is spec compliant
+// and is followed by a ":". After that anything is accepted.
+//
+// Reference: https://url.spec.whatwg.org/#url-scheme-string.
+const URL_VALIDATION_REGEX = /^[a-zA-Z][a-zA-Z0-9-\+\.]*:(.*)$/;
+
+/**
+ * Error thrown when there is a URL validation error that also yields error recording.
+ */
+class UrlMetricError extends Error {
+  constructor(readonly type: ErrorType, message?: string) {
+    super(message);
+    this.name = "UrlMetricError";
+  }
+}
+
+export class UrlMetric extends Metric<string, string> {
+  constructor(v: unknown) {
+    super(v);
+  }
+
+  /**
+   * Validates that a given value is a valid URL metric value.
+   *
+   * 1. The URL must be a string.
+   * 2. The URL must have a maximum length of URL_MAX_LENGTH characters before the encoding.
+   * 3. The URL must not be a data URL.
+   * 4. Every URL must start with a valid scheme.
+   *
+   * Note**: We explicitly to not validate if the URL is fully spec compliant,
```

Comment:
```suggestion
   * **Note**: We explicitly do not validate if the URL is fully spec compliant,
```

I assume this uses markdown syntax?

Reply:
Yeah, the jsdoc to HTML thingy we use will transform this in HTML.